### ISSUE TITLE
intl402: Intl/getCanonicalLocales/main.js - islamicc is deprecated and need to be canonicalized to ismalic-civil? 

### ISSUE THREAD

**Issue by jungshik** [tc39/test262#743]
intl402/Intl/getCanonicalLocales/main.js has the following: 

```
assertArray(gCL(['ar-ma-u-ca-islamicc']), ['ar-MA-u-ca-islamicc']);
```

v8's implementation (that uses two ICU APIs in series: uloc_fromLanguage() and uloc_toLanguage()) replacens 'islamicc' (deprecated in LDML) with 'islamic-civil' (its preferred form).  See 
http://www.unicode.org/repos/cldr/tags/latest/common/bcp47/calendar.xml

 Ecma 402 6.2.3 (
https://tc39.github.io/ecma402/#sec-canonicalizelanguagetag ) has the following:

> It takes the steps specified in RFC 5646 section 4.5, or successor, to bring the language tag into 
> canonical form, and to regularize the case of the subtags, but does not take the steps to bring a 
> language tag into "extlang form" and to reorder variant subtags.

RFC 5646 section 4.5 has the following:

>  point 2.  Redundant or grandfathered tags are replaced by their 'Preferred-
>       Value', if there is one.
> 
>  point 3.  Subtags are replaced by their 'Preferred-Value', if there is one.
>    For extlangs, the original primary language subtag is also
>    replaced if there is a primary language subtag in the 'Preferred-
>    Value'.

RFC 6067 for the '-u-' extension does not say anything special about deprecated vs preferred. 

My interpretation is that '-u-ca-islamicc' has to be replaced by '-u-ca-islamic-civil'  (btw, this is spun off from https://bugs.chromium.org/p/v8/issues/detail?id=5012 )

@zbraniecki ,  what would you say?

**Comment by jungshik** [comment]
Hmm... it seems like I need to dig a bit more into this issue.

**Comment by jungshik** [comment]
What ICU's uloc_forLanguageTag (input bcp47 tag => ICU's internal locale id) and uloc_toLanguageTag (ICU's internal locale id => canonicalized BCP 47 tag) do is:

```
ar-MA-u-ca-islamicc	ar_MA@calendar=islamic-civil	ar-MA-u-ca-islamic-civil
ar-MA-u-ca-islamic-civil	ar_MA@calendar=islamic-civil	ar-MA-u-ca-islamic-civil
```

The first column is the input, the second is ICU's internal locale id and the 3rd is the canonicalized output. Note that 'islamicc' (deprecated) is replaced by 'islamic-civil' (preferred).

**Comment by zbraniecki** [comment]
@jswalden Waldo, can you take a look at this? It seems to me that we use our own canonicalization code in https://dxr.mozilla.org/mozilla-central/rev/0502bd9e025edde29777ba1de4280f9b52af4663/js/src/builtin/Intl.js#362 and we could replace it with ICU like Google does.

This would reduce the size of the Intl.js file and make us compatible with what @jungshik is requesting.

Does it make sense?

**Comment by zbraniecki** [comment]
From https://bugzilla.mozilla.org/show_bug.cgi?id=837961#c30 -

> the test262 commit is wrong, because ECMA402 doesn't require to canonicalize language extension sequences [1]:
> 
> The specifications for extensions to BCP 47 language tags, such as RFC 6067, may include canonicalization rules for the extension subtag sequences they define that go beyond the canonicalization rules of RFC 5646 section 4.5. Implementations are allowed, but not required, to apply these additional rules. 
>
>
> RFC 5646 doesn't specify Unicode extension sequences, so it doesn't make sense to refer to RFC 5646 section 4.5 in the test262 commit. Furthermore Unicode extension sequence canonicalization (RFC 6067, 2.1.1 [2]) doesn't specify that subtags within a keyword are updated when deprecated values are used. So I'd say the test262 commit is invalid and should be reverted.
> 
> 
> [1] https://tc39.github.io/ecma402/#sec-canonicalizelanguagetag
> [2] https://tools.ietf.org/html/rfc6067#section-2.1.1

@jungshik - would you agree with this and if so, can you revert your change, please?

**Comment by zbraniecki** [comment]
@tcare  - can I get an update on this please? I believe this should be reverted.

**Comment by leobalter** [comment]
this may be closed for now?

**Comment by jungshik** [comment]
At minimum,  I don't think that  https://github.com/tc39/test262/issues/743#issuecomment-252066935 is not for reverting the change but for removing the test.

**Comment by jungshik** [comment]
I can't reopoen this issue. Can somebody reopen ?

**Comment by zbraniecki** [comment]
@jungshik - do you feel that we should remove this test instead of unifying engine behavior?

**Comment by leobalter** [comment]
@zbraniecki based on https://github.com/tc39/test262/pull/774#issuecomment-254534412, this test needs to be removed.

The "not required" part means we can't assert one way or another.

**Comment by littledan** [comment]
@zbraniecki Let's separate getting the tests aligned with the current spec from fixing the spec.

**Comment by jungshik** [comment]
@zbraniecki 
RFC 6067 says rather little about the canonicalization (implicitly) deferring to UTR 35, which has the following: 

```
preferred
The preferred value of the deprecated key, type or attribute element. When a key, type or attribute element is deprecated, this attribute is used for specifying a new canonical form if available.

alias (Not applicable to <attribute>)
The BCP47 form is the canonical form, and recommended. Other aliases are included only for backwards compatibility.

Example:
<type name="phonebk" alias="phonebook" description="Phonebook style ordering (such as in German)"/>
The preferred term, and the only one to be used in BCP47, is the name: in this example, "phonebk".
The alias is a key or type name used by Unicode locale extensions with the old "@key=type" syntax. The attribute value for type may contain multiple names delimited by ASCII space characters. Of those aliases, the first name is the preferred value.
```

1. Given above, I'd argue that canonicalzing ca-islamicc to 'ca-islamic-civil' is per RFC 6067. 

2. what RFC 5646 section 4.5 requires is up for the interpretation. It can be certainly interpreted that a 'preferred-value' provision is *not* applicable to '-u-' extension subtags because the provision only talks about non-extension (sub)tags. 

3. https://tc39.github.io/ecma402/#sec-canonicalizelanguagetag has the following. 

```
The specifications for extensions to BCP 47 language tags, such as RFC 6067, may include canonicalization rules for the extension subtag sequences they define that go beyond the canonicalization rules of RFC 5646 section 4.5. Implementations are allowed, but not required, to apply these additional rules.
```

If we follow the interpretation of RFC 5646 section 4.5 in "2." above and combine that with "3." and "1.", we have to remove the test for 'ca-islamicc'. 

If we follow the interpretation of RFC 5646 section 4.5 in "2." above and combine that with "3." but NOT "1.",  we have to keep the latest change (reverting my change).  However, I believe that "1." is the case (RFC 6067 has to be interpreted in combination with UTR 35). 

My interpretation of RFC 5646 section 4.5 is different from "2.", which is why I came up with that change.

**Comment by littledan** [comment]
@jungshik Your explanation of 1. makes of sense to me. Let's delete this test and work on clarifying the wording in the spec in https://github.com/tc39/ecma402/issues/111.

**Comment by littledan** [comment]
Further support for @jungshik 's interpretation: in [Appendix A](https://tc39.github.io/ecma402/#annex-implementation-dependent-behaviour) of ECMA 402, the second bullet of the first bullet (why are these unordered lists?) mentions "Canonicalization of extension subtag sequences beyond the rules of RFC 5646 (6.2.3)" as an implementation-dependent extension.

**Comment by leobalter** [comment]
this should be fixed by #789. Please reopen if necessary.
